Feature #27: Repair Terran buildings

ImpulseCloud opened the pull request:
When a building is first damaged, sends one repairer.
If no current repairers, sends one repairer to first still-damaged building.

alkurbatov commented on src/Dispatcher.cpp:
```diff
@@ -125,6 +125,13 @@ void Dispatcher::OnUnitIdle(const sc2::Unit* unit_) {
 }
 
+void Dispatcher::OnUnitDamaged(const sc2::Unit* unit_, float health_, float shields_) {
+    gHub->OnUnitDamaged(*unit_, health_, shields_);
```
Why do we need to notify Hub about this event?

ImpulseCloud replied:
Mostly just to go with the common 'standard' of notifying Hub and then all the plugins for any event. I can remove this from Hub though.

alkurbatov commented on src/Hub.cpp:
```diff
@@ -271,6 +274,16 @@ sc2::Tag Hub::AssignBuildTask(Order* order_, const sc2::Point2D& point_) {
 
+bool Hub::AssignRepairTask(const sc2::Unit& damaged_) {
+    const sc2::Point2D repair_location(damaged_.pos);
+    Worker* worker = GetClosestFreeWorker(repair_location);
```
I think the Repairman plugin should do this action, especially since you don't try to mark this worker as busy which might be a bug.

ImpulseCloud replied:
'GetClosestFreeWorker' marks the worker as 'busy' before it returns the worker.  (I don't think it really should do that, but that is its current existing behavior. I'd rather a new AssignedTask should be doing the 'busy' marking.)

The reason I have it in Hub, is so it can be called by Strategy to micro Combat Repair of Mechanical Units. We don't yet have a way for Strategy to talk to another Plugin. Even a 'Strategy Settings' struct would not allow Strategy to dynamically call RepairMan::AssignRepairTask for damaged military units. I really do think Hub should have a pointer to every other plugin, like CommandCenter does.

Hub contains the free/busy_workers list, so I feel that Hub should manage the actual job assignment just like 'AssignRefineryConstruction', 'AssignBuildTask', 'AssignVespeneHarvester'. The Repairman contains the logic on when it should happen.

alkurbatov replied:
According to my design Hub is a cache. It contains a picture of the real world with all these lists and additional info the bot needs. All the game logic should be moved out of it and the hub itself should be simplified a lot (e.g. all that different lists should gone).
The subsystems act independently until a special messaging channel is set, such approach allows to make very agile system with clear division between parts. Also, the subsystems divided by purpose intentionally, it allows to keep related code inside. The idea about single point of taking decision and executing them immediately could work but not in this architecture.

So, I am don't like the idea of creating a command center out of Hub.

ImpulseCloud replied:
Okay, moved AssignRepairTask to Repairman.

alkurbatov commented on src/main.cpp:
```diff
@@ -33,7 +33,7 @@ int main(int argc, char* argv[]) {
     Dispatcher bot("TrainingDummy");
     coordinator.SetParticipants({
-        CreateParticipant(sc2::Race::Random, &bot, "Suvorov"),
+        CreateParticipant(sc2::Race::Terran, &bot, "Suvorov"),
```
Please, revert this change.

ImpulseCloud replied:
Done. (forgot to remove after testing, not intended)

I think we should change the default AI skill level to Hard or VeryHard though, as that is about the level I think the current bot builds have a chance against. Thoughts? For new authors, it is not very motivating to try out and get absolutely destroyed by the AI.

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -2,15 +2,57 @@
+
 void RepairMan::OnStep(Builder*) {
-    if (gAPI->observer().GetCurrentRace() != sc2::Race::Terran)
+    if (gHub->GetCurrentRace() != sc2::Race::Terran || m_damaged_buildings.empty())
```
There is no need in checking current race as this plugin should be activated only for Terran. See the initialization code in Dispatcher.

alkurbatov replied:
Ah I am wrong here, this plugin used to recreate destroyed buildings, too. So tis part I ok.

ImpulseCloud replied:
Yes, I got this wrong too at first. I changed the plugin to be activated only for Terrans, but later I saw the recreate-destroyed-buildings logic and had to change it all back. Probably, the recreate-destroyed-buildings should be in Builder or something, as it will likely want to consider current needs rather than just auto-recreate, if it isn't a needed tech building. But since I don't have a plan for that yet, just keeping it as it is currently.

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -2,15 +2,57 @@
+
+void RepairMan::OnUnitDamaged(const sc2::Unit* unit_,
+                             float health_, float shields_, Builder*) {
+    if (gHub->GetCurrentRace() != sc2::Race::Terran || IsCombatUnit()(*unit_))
```
You can use the new `IsBuilding` filter from the API instead of `IsCombatUnit`. Just to make the code a bit more self-descriptive. It also allows to avoid `m_ignoreWorkers`.
Of course we may need to repair workers sometimes, but that should be in units micro.

ImpulseCloud replied:
Done.

alkurbatov commented on src/plugins/RepairMan.h:
```diff
@@ -10,5 +10,11 @@
+
+ private:
+    bool m_ignoreWorkers = true;
+    sc2::Units m_damaged_buildings;
```
Should be list of sc2::Unit as you frequently remove repaired buildings from it.

ImpulseCloud replied:
Done.

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -2,15 +2,57 @@
+    auto it = m_damaged_buildings.begin();
+    while (it != m_damaged_buildings.end()) {
+        if ((*it)->health == (*it)->health_max) {
+            gHistory.info() << sc2::UnitTypeToName((*it)->unit_type) <<
```
Please, use `std::erase_if` with e.g. lambda for vector or embedded `remove_if` for list.

ImpulseCloud replied:
Done.

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -2,15 +2,57 @@
+
+    auto repairers = gAPI->observer().GetUnits(IsRepairer());
+    if (!m_damaged_buildings.empty() && repairers().empty()) {
+        gHub->AssignRepairTask(*m_damaged_buildings.front());
```
What happens with the assigned SCVs when a building is repaired?

ImpulseCloud replied:
SCV becomes idle, and Hub::OnIdle turns it 'free', then Miner::OnIdle calls DistributeWorker to send to nearest base to gather minerals. If the building is attacked again, it will still be close and will win the GetClosestFreeWorker and go to repair again.

For now, this avoids extra logic of staying near the building until the enemy leaves, but that can be done in the future for advanced Combat Repair micro.

alkurbatov reviewed (changes requested):
Checked, several changes required.

I'll check other requests when we finish with this one as the diff has common parts between the requests.

alkurbatov commented:
A little style comment: please, use the same approach in the commits naming as I did (e.g. `feat: Some feature, #id`). Please, take a look at the commit's history for additional info.

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -2,15 +2,48 @@
         return;
 
-    // FIXME (alkuratov): Put buildings repair code here.
+    auto it = std::remove_if(m_damaged_buildings.begin(), m_damaged_buildings.end(),
```
It would be better to use `std::erase_if` as it'll immediately clean up the list. Sorry, originally I suggested to use `std::remove_if`, but later changed the comment. Maybe that passed unnoticed.

alkurbatov replied:
Also, `std::list` has embedded `remove_if` which works much faster comparing to the general version.

ImpulseCloud replied:
Awesome! Changed to embedded std::list::remove_if, which does the erase too.

"This calls the destructor of these objects and reduces the container size by the number of elements removed."
https://www.cplusplus.com/reference/list/list/remove_if/

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -2,15 +2,48 @@
+
 void RepairMan::OnStep(Builder*) {
-    if (gAPI->observer().GetCurrentRace() != sc2::Race::Terran)
+    if (gHub->GetCurrentRace() != sc2::Race::Terran || m_damaged_buildings.empty())
```
This `m_damaged_buildings.empty()` check is not needed as below we have another one right after `std::remove_if`. If removal done in one call, it'll be safe to check the list for emptiness later.

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -2,15 +2,48 @@
+void RepairMan::OnUnitDamaged(const sc2::Unit* unit_,
+                             float health_, float shields_, Builder*) {
+    if (gHub->GetCurrentRace() != sc2::Race::Terran || !sc2::IsBuilding()(*unit_))
+        return;  // not applicable
```
Style: please, drop obvious comments.

ImpulseCloud reviewed (commented):
Requested changes made.

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -2,15 +2,43 @@
-    // FIXME (alkuratov): Put buildings repair code here.
+    m_damaged_buildings.remove_if( [](const sc2::Unit* unit_) {
+            return unit_->health == unit_->health_max || !unit_->is_alive;
+        });
```
Style:  Too deep indent.

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -2,15 +2,43 @@
+
+    auto repairers = gAPI->observer().GetUnits(IsRepairer());
+    if (repairers().empty() && !m_damaged_buildings.empty()) {
+        AssignRepairTask(*m_damaged_buildings.front());
```
Style: Extra braces.

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -2,15 +2,43 @@
+        });
+
+    auto repairers = gAPI->observer().GetUnits(IsRepairer());
+    if (repairers().empty() && !m_damaged_buildings.empty()) {
```
So it is like one repairer at a time? Shouldn't it be larger count, e.g. one repairer per unit?

alkurbatov replied:
Also, from performance point of view, it'll be better to bail out on `m_damaged_buildings.empty()` and avoid that heavy GetUnits call.

ImpulseCloud replied:
So, when I tested, an enemy army came and damaged like ~5 buildings near ramp and 5 separate workers come to repair the buildings, and they all get killed and then another 5 come and get killed, etc, which quickly killed all my workers. So keeping it to 1 was an attempt to reduce massive worker-deaths.

That's *after* the initial repair attempt though. In OnDamaged, each first-time-damaged building gets a repairer, but if it dies, another doesn't get sent unless there are no repairers currently, at which point it will send a repairer to the building that was damaged first.

I think 3 repairers minimum is a better balance, but wanted to start at 1 at least (because it was easier to code). I don't want to put in code that creates a worse behavior than not repairing at all, and having a continual line of many workers going to their death seems to be *worse* than no repairing (as it takes a while for enemy to kill buildings, and lets us build more military in the meantime).

For *advanced* repair behavior, it needs to take into account where enemies are, but that need to be done in Marine Micro, or elsewhere first, and is not available yet. I wanted to do a 'minimum' useful Repair Behavior first. After Secure/Defend-Expansion Feature and Marine-Micro feature, the defending army should respond to building damage and protect the repairing workers. After that, we could increase the number of repairers.

Another option (for now) is to send the Repairer away if it gets to <50% of its own health.

alkurbatov replied:
3 repairers looks good to me. 
Agree on <50% health.

ImpulseCloud replied:
Ran into a funny bug. Forgot that OnDamaged is called for Enemy units too. So when I attacked an Enemy Terran Turret, it got into m_damaged_buildings, and tried to command self SCVs to Repair, but the command would fail, so SCV would stay on GATHER_MINERALS, but still be in m_busy_workers, and there would be no Repairers, so it would try another SCV, until all other SCV's were in m_busy_workers as REPAIRING.

ImpulseCloud replied:
Same bug effect also occurred when Minerals = 0, REPAIR command fails and GATHER_MINERALS never stops. Fixed by checking minerals before assigning REPAIR, though this doesn't prevent a sudden construction cost from dropping minerals to 0 on next frame. Fixed that by first command STOP, then queueing REPAIR, so even if REPAIR fails, worker is stopped from GATHER_MINERALS and will then call OnIdle properly.

ImpulseCloud replied:
Adding the "<50%" took more additions than I thought. Had to add filter functions to GetClosest methods to prevent trying to grab the same damaged worker that I was sending away.

alkurbatov commented on src/plugins/RepairMan.h:
```diff
@@ -10,5 +10,12 @@
+
     void OnUnitDestroyed(const sc2::Unit* unit_, Builder* builder_) final;
+
+    bool AssignRepairTask(const sc2::Unit& damaged_);
```
Should be private as this is very internal method, not used anyhow outside.

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -53,3 +81,13 @@ void RepairMan::OnUnitDestroyed(const sc2::Unit* unit_, Builder* builder_) {
+        return false;
+
+    worker->Repair(damaged_);
+    return true;
```
Seems that the returned value is not used anyhow and this method should be void.

alkurbatov commented on src/plugins/RepairMan.h:
```diff
@@ -10,5 +10,12 @@
+    bool AssignRepairTask(const sc2::Unit& damaged_);
+
+ private:
+    std::list<const sc2::Unit*> m_damaged_buildings;
```
Style: Please, include the corresponding header.

ImpulseCloud replied:
What do you mean here? I'm not certain what "corresponding header" refers to about line 20.

for std::list , I should include a list.h ?
for Unit*, I should include a sc2 unit.h ?

ImpulseCloud replied:
Please provide more info here. I am confused.

alkurbatov replied:
According to the Google styleguide, we should include std headers (`<list>` here) explicitly, even if the required header is available from another header.
Technically they recommend to include every required header (so unit.h might be needed, too) in order to avoid compilation problems if some other included headers change.

alkurbatov reviewed (changes requested):
Thank you! Minor changes still required.

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -53,3 +95,18 @@ void RepairMan::OnUnitDestroyed(const sc2::Unit* unit_, Builder* builder_) {
+void RepairMan::AssignRepairTask(const sc2::Unit& damaged_) {
+    std::function<bool(const Worker&)> moreThanHalfHealth =
+    [](const Worker& worker) {
+        const sc2::Unit* unit = gAPI->observer().GetUnit(worker.Tag());
```
This looks strange. Why do you try to check health of a worker? I though we agreed on fixing building with health <= 50%.

ImpulseCloud replied:
Had to add filter functions to GetClosest methods to prevent trying to grab the same damaged worker that I was sending away. If I send away a worker to minerals with health <=50% and Free them, the GetClosestFree will just re-grab the same worker (or possibly another worker <=50% health). So the <=50% part doesn't work without the filter.

ImpulseCloud replied:
Oh, I see now. You thought I meant only repair when the Building gets below 50%. I meant retreating the Worker when worker.health<=50% so it will not die. I can add it so Buildings don't get repairers until they are <=50% so it doesn't send 3 workers when a building gets only a single hit of damage.

alkurbatov replied:
Well, it is up to you. I would ignore the buildings until some point (e.g. 50 or 75 percent of damage, 75 could be even better), but repairing after first damage event makes sense, too.

I would totally ignore worker health as this can be easily hardcoded in unit's micro for a more specific way (e.g. ask a worker to run away if it's hp is too low).  Also, it allows to force worker to retreat in other situations and can be generalised for other units.

ImpulseCloud replied:
Ok, I'll ignore Worker health for now and later in Unit micro, create RETREATING job behavior.

alkurbatov commented on src/plugins/RepairMan.cpp:
```diff
@@ -53,3 +95,18 @@ void RepairMan::OnUnitDestroyed(const sc2::Unit* unit_, Builder* builder_) {
+    };
+
+    const sc2::Point2D repair_location(damaged_.pos);
+    Worker* worker = gHub->GetClosestFreeWorker(repair_location, moreThanHalfHealth);
```
If such filtering really needed, I suggest to do it in a more functional way, i.e. get list of workers with required health value. Then filter it by distance.

ImpulseCloud replied:
Do you mean:
Get const ref of Cache<Worker> m_free_workers (Have to add Hub::GetFreeWorkers() for this).
Iterate through m_free_workers and put Filtered workers in a temp Cache<Worker>.
Then use Cache::GetClosestTo on the filtered Cache

I believed you would want me to *avoid* creating copies of the Workers. 

If I instead filter to a list of Worker*, I can't use the Cache::GetClosestTo and have to do a manual distance check. Is this what you are suggesting?

To me, passing in a filter is the simple and flexible way to achieve the goal. As it's passing a function as a first-order object, would it not be considered the more "functional way"?

alkurbatov replied:
First of all, I don't like the current implementation of GetClosestFreeWorker as it always checks do we have functor or not. It means that technically we have two different methods wrapped inside one. The functional way allows to solve this task in a more elegant way, i.e.

a. Get free workers list.
b. Create new list of healthy workers.
c. Find the closest one.

As a result at least one copy of units array will be created, but it'll not provide large overhead I think. Also, we can mitigate it by using std::move, although I can't give particular advice on it as sadly I am not familiar with this new mechanic.

Another simple improvement could be inclusion of pointer to sc2::Unit inside GameObject class. The original design assumed that usually we don't need the pointer but can use the current task most of the time. Now it is not enough so we can avoid GetUnit by tag call with this small fix. The API guaranties that the pointer will be kept alive whole game.

ImpulseCloud replied:
Originally, I did create GetClosetFreeWorker / GetClosestTo Filtered versions as separate additional functions, but then I thought you would not like as it is the exact same code with just 2 extra checks, so I combined them.

I'll make filtered copy array and add Unit*.

ImpulseCloud replied:
I see the GameObject has two ToUnit functions which return a dummy unit with only a valid Tag, for sending to API. I'm suggesting to rename these to 'ToDummyUnit' so they are not confused with a Unit/Unit* to a real unit with full data. (I was confused and had to figure that out earlier)

alkurbatov replied:
The only reason `ToUnit` exists is the API which doesn't support command execution by tag. That should be fixed on the API's side and `ToUnit` should be removed. Initially I was not sure about the way API handles all that pointers so it looked (and it still looks) not safe to keep the pointer. However, the internals says that the pointer stays valid until end of a game.

However, the simplest solution: make the unit pointer public without additional getters.

ImpulseCloud replied:
@alkurbatov  I looked up std::move, and it is for used to move a temporary object (rvalue reference T&&) into a permanent variable without copy-constructor, but it sets the old object to null, so if used on m_free_workers, it would zero out the object.

https://stackoverflow.com/questions/3413470/what-is-stdmove-and-when-should-it-be-used
"It's a new C++ way to avoid copies. For example, using a move constructor, a std::vector could just copy its internal pointer to data to the new object, leaving the moved object in a moved-from state, therefore not copying all the data. This would be C++-valid."

https://en.wikipedia.org/wiki/C%2B%2B11#Rvalue_references_and_move_constructors
"In C++11, a move constructor of std::vector<T> that takes an rvalue reference to an std::vector<T> can copy the pointer to the internal C-style array out of the rvalue into the new std::vector<T>, then set the pointer inside the rvalue to null. Since the temporary will never again be used, no code will try to access the null pointer, and because the pointer is null, its memory is not deleted when it goes out of scope."

alkurbatov replied:
You need to create a new object after filtration of m_free_workers i.e. a vector with > 50% hp workers. This tmp object could be returned to caller with std::move. It allows to avoid doing 1 on 2 copies.

alkurbatov replied:
Ok, after digging inside std::move I should say that we can't use it to optimise return of local values like filtered arrays :(

ImpulseCloud committed:
feat: Repair Terran buildings, #27

When a building is first damaged, sends one repairer.
OnStep, if less than 3 total repairers, sends an additional repairer to first still-damaged building.

ImpulseCloud commented:
I removed the <50% worker-health logic, and will save that for unit micro, as you suggest. I've saved my code/notes for the future.

For Building health percentage, I'll use "repairing after first damage event makes sense, too." for now.

I'll save "Unit* in GameObject" for the next PR I use it in.

alkurbatov commented:
Thank you! LGTM.
